File: GoClient/rpc_client.go
package main

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/pebbe/zmq4"
)

type RPCPubSubClient struct {
	rpcAddr string
	subAddr string

	req *zmq4.Socket
	sub *zmq4.Socket

	callbacks map[string]func(string, string)
	mu        sync.Mutex

	running  bool
	stopChan chan struct{}
	wg       sync.WaitGroup
}

func NewRPCPubSubClient(name string) (*RPCPubSubClient, error) {
	client := &RPCPubSubClient{
		rpcAddr:   "ipc:///tmp/" + name + "_rpc.sock",
		subAddr:   "ipc:///tmp/" + name + "_pub.sock",
		callbacks: make(map[string]func(string, string)),
		running:   true,
		stopChan:  make(chan struct{}),
	}

	var err error

	client.req, err = zmq4.NewSocket(zmq4.REQ)
	if err != nil {
		return nil, fmt.Errorf("failed to create REQ socket: %w", err)
	}
	if err = client.req.Connect(client.rpcAddr); err != nil {
		client.req.Close()
		return nil, fmt.Errorf("failed to connect REQ to %s: %w", client.rpcAddr, err)
	}

	client.sub, err = zmq4.NewSocket(zmq4.SUB)
	if err != nil {
		client.req.Close()
		return nil, fmt.Errorf("failed to create SUB socket: %w", err)
	}

	if err = client.sub.Connect(client.subAddr); err != nil {
		client.req.Close()
		client.sub.Close()
		return nil, fmt.Errorf("failed to connect SUB to %s: %w", client.subAddr, err)
	}

	client.wg.Add(1)
	go client.listenSub()

	return client, nil
}

func (c *RPCPubSubClient) Call(timeoutMs int, method string, params ...interface{}) (interface{}, error) {
	// Set timeout
	c.req.SetRcvtimeo(time.Duration(timeoutMs) * time.Millisecond)

	request := map[string]interface{}{
		"method": method,
		"params": params,
	}

	reqData, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	if _, err := c.req.SendBytes(reqData, 0); err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}

	respData, err := c.req.RecvBytes(0)
	if err != nil {
		return nil, fmt.Errorf("RPC call '%s' timed out: %w", method, err)
	}

	var response map[string]interface{}
	if err := json.Unmarshal(respData, &response); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	if response["error"] != nil {
        if errMsg, ok := response["error"].(string); ok {
            if errMsg == "" {
                return nil, fmt.Errorf("RPC error: empty error message")
            }
            return nil, fmt.Errorf("RPC error: %s", errMsg)
        }
        return nil, fmt.Errorf("RPC error: invalid error type")
	}

	return response["result"], nil
}

func (c *RPCPubSubClient) Subscribe(topic string, callback func(string, string)) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if err := c.sub.SetSubscribe(topic); err != nil {
		return fmt.Errorf("failed to subscribe to %s: %w", topic, err)
	}

	c.callbacks[topic] = callback
	return nil
}

func (c *RPCPubSubClient) Unsubscribe(topic string) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if err := c.sub.SetUnsubscribe(topic); err != nil {
		return fmt.Errorf("failed to unsubscribe from %s: %w", topic, err)
	}

	delete(c.callbacks, topic)
	return nil
}

func (c *RPCPubSubClient) listenSub() {
	defer c.wg.Done()

	for c.running {
		select {
		case <-c.stopChan:
			return
		default:
			msg, err := c.sub.RecvMessage(zmq4.DONTWAIT)
			if err != nil {
				time.Sleep(10 * time.Millisecond)
				continue
			}

			if len(msg) < 2 {
				continue
			}

			topic := msg[0]
			payload := msg[1]

			c.mu.Lock()
			callback, exists := c.callbacks[topic]
			c.mu.Unlock()

			if exists && callback != nil {
				callback(topic, payload)
			}
		}
	}
}

func (c *RPCPubSubClient) Stop() {
	c.running = false
	close(c.stopChan)
	c.wg.Wait()

	if c.req != nil {
		c.req.Close()
	}
	if c.sub != nil {
		c.sub.Close()
	}
}

File: PythonClient/rpc_client.py
import zmq
import threading
import json
import time

class RPCPubSubClient:
    def __init__(self, name):
        self.context = zmq.Context()

        rpc_addr = "ipc:///tmp/" + name + "_rpc.sock"
        sub_addr = "ipc:///tmp/" + name + "_pub.sock"

        self.req = self.context.socket(zmq.REQ)
        self.req.connect(rpc_addr)

        self.sub = self.context.socket(zmq.SUB)
        self.sub.connect(sub_addr)

        self._callbacks = {}
        self._lock = threading.Lock()

        self._running = True

        self._sub_thread = threading.Thread(target=self._listen_sub, daemon=True)
        self._sub_thread.start()

    def call(self, method, *params, timeout_ms=3000):
        """Wywołanie RPC synchroniczne z timeoutem w ms"""
        self.req.setsockopt(zmq.RCVTIMEO, timeout_ms)

        req = {"method": method, "params": params}
        self.req.send_json(req)
        try:
            resp = self.req.recv_json()
        except zmq.Again:
            raise TimeoutError(f"RPC call '{method}' timed out after {timeout_ms} ms")

        # Check errorCode (0 = success)
        error_code = resp.get("errorCode", 0)
        if error_code != 0:
            error_msg = resp.get("errorMsg", "unknown error")
            raise Exception(f"RPC error (code {error_code}): {error_msg}")

        return resp.get("result")

    def __getattr__(self, name):
        def method(*args, **kwargs):
            # kwargs mogą zawierać timeout_ms
            return self.call(name, *args, **kwargs)
        return method

    def subscribe(self, topic, callback):
        self.sub.setsockopt_string(zmq.SUBSCRIBE, topic)
        with self._lock:
            self._callbacks[topic] = callback

    def unsubscribe(self, topic):
        self.sub.setsockopt_string(zmq.UNSUBSCRIBE, topic)
        with self._lock:
            if topic in self._callbacks:
                del self._callbacks[topic]

    def stop(self):
        """Zatrzymanie wątku SUB i zamknięcie socketów"""
        self._running = False
        self._sub_thread.join(timeout=1)
        self.sub.close()
        self.req.close()
        self.context.term()

    def _listen_sub(self):
        while self._running:
            try:
                parts = self.sub.recv_multipart(flags=zmq.NOBLOCK)
            except zmq.Again:
                time.sleep(0.01)
                continue

            topic = parts[0].decode()
            payload = parts[1].decode() if len(parts) > 1 else ""
            try:
                payload = json.loads(payload) if payload else {}
            except json.JSONDecodeError:
                payload = {"raw": payload}

            with self._lock:
                cb = self._callbacks.get(topic)
            if cb:
                cb(topic, payload)


if __name__ == "__main__":
    def on_time(topic, data):
        print("[TIME]", data)

    def on_weather(topic, data):
        print("[WEATHER]", data)

    client = RPCPubSubClient("test")
    client.subscribe("time", on_time)
    client.subscribe("weather", on_weather)

    result = client.add(5,3)
    print(f"add(5,3) = {result}")
    result = client.multiply(4,2)
    print(f"multiply(4,2) = {result}")

    time.sleep(10)
    print("Unsubscribing from 'weather'")
    client.unsubscribe("weather")

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("Stopping client...")
        client.stop()
